src/MobileRedirection.php

```diff
@@ -0,0 +1,530 @@
+<?php
+/**
+ * Class MobileRedirection.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP;
+
+use AMP_Options_Manager;
+use AMP_Theme_Support;
+use AmpProject\AmpWP\Infrastructure\Registerable;
+use AmpProject\AmpWP\Infrastructure\Service;
+use AmpProject\Attribute;
+
+/**
+ * Service for redirecting mobile users to the AMP version of a page.
+ *
+ * @package AmpProject\AmpWP
+ */
+final class MobileRedirection implements Service, Registerable {
+
+	/**
+	 * Regular expression for regular expressions. So meta.
+	 *
+	 * This must work in both PHP and cross-browser JS, which is why the 's' flag is not used. Also, this will get
+	 * passed as the pattern argument to the RegExp constructor in JS, whereas in PHP it will be used as the pattern
+	 * surrounded by the '#' delimiter.
+	 *
+	 * @var string
+	 */
+	const REGEX_REGEX = '^\/((?:.|\n)+)\/([i]*)$';
+
+	/**
+	 * The name of the cookie that persists the user's preference for viewing the non-AMP version of a page when on mobile.
+	 *
+	 * @var string
+	 */
+	const DISABLED_COOKIE_NAME = 'amp_mobile_redirect_disabled';
+
+	/**
+	 * Query parameter to indicate that the page in question should not be served as AMP.
+	 *
+	 * @todo This needs to not be specific to mobile.
+	 *
+	 * @var string
+	 */
+	const NO_AMP_QUERY_VAR = 'noamp';
```

Comment:
Given that `AMP_Validation_Manager` also needs to redirect to the non-AMP version with a `noamp` query param, perhaps this should not be located in this service.

Reply:
I'd vote for a `QueryVar` interface (or class if logic is needed) with this and other constants.